### PR TITLE
feat[CL]: add asset breakdown to positions

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #4632 

## What is the purpose of the change

When querying position, the only thing that was returned was a position struct. While this has useful information, the underlying amount of asset0 and asset1 was not returned. This change returns both the asset0 and asset1 breakdown for each user position when calling the user-positions query.

## Brief Changelog

- Adds `PositionWithUnderlyingAssetBreakdown` proto struct
  - This is essentially a model.Position as it was before, but adds asset0 and asset1 sdk.Dec fields
- Implements this proto struct in the `user-positions` query  

## Testing and Verifying

I spun up a local osmosis instance with CL state and queried it. Received the following:

```
positions:
- asset0: "5000000000.000000000000000000"
  asset1: "999999.929289321953752005"
  position:
    address: osmo1jllfytsz4dryxhz5tl7u73v29exsf80vz52ucc
    freeze_duration: 0s
    join_time: "2023-03-15T23:34:18.294912471Z"
    liquidity: "70710678.118654755000100000"
    lower_tick: "-1620000"
    pool_id: "4"
    upper_tick: "3420000"
- asset0: "5000000000.000000000000052404"
  asset1: "0.000000000000000000"
  position:
    address: osmo1jllfytsz4dryxhz5tl7u73v29exsf80vz52ucc
    freeze_duration: 0s
    join_time: "2023-03-15T23:34:33.429215298Z"
    liquidity: "2997892628892.147707843478961141"
    lower_tick: "300000"
    pool_id: "4"
    upper_tick: "309990"
- asset0: "4999999999.999999999999945060"
  asset1: "0.000000000000000000"
  position:
    address: osmo1jllfytsz4dryxhz5tl7u73v29exsf80vz52ucc
    freeze_duration: 0s
    join_time: "2023-03-15T23:34:23.335370459Z"
    liquidity: "3706060757243.600559263910088981"
    lower_tick: "305450"
    pool_id: "4"
    upper_tick: "315000"
- asset0: "5000000000.000000000000067028"
  asset1: "0.000000000000000000"
  position:
    address: osmo1jllfytsz4dryxhz5tl7u73v29exsf80vz52ucc
    freeze_duration: 0s
    join_time: "2023-03-15T23:34:28.384895378Z"
    liquidity: "5988837224777.540616115005684301"
    lower_tick: "315000"
    pool_id: "4"
    upper_tick: "322500"
```
## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/concentrated-liquidity/grpc_query.go [hunk 1]
```diff
@@ -77,8 +78,37 @@ func (q Querier) UserPositions(ctx context.Context, req *clquery.QueryUserPositi
...
+		pool, err := q.Keeper.GetPool(sdkCtx, position.PoolId)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		// transform poolI to concentratedPoolExtension
+		concentratedPoolExtension, ok := pool.(types.ConcentratedPoolExtension)
+		if !ok {
+			return nil, status.Error(codes.Internal, "pool is not concentrated pool")
+		}
```
Comment (p0mvn) [inline x/concentrated-liquidity/grpc_query.go:94]: There is `GetPoolById` that can get the right type directly

File: x/concentrated-liquidity/grpc_query.go [hunk 2]
```diff
@@ -77,8 +78,37 @@ func (q Querier) UserPositions(ctx context.Context, req *clquery.QueryUserPositi
...
+
+		// Transform the provided ticks into their corresponding sqrtPrices.
+		sqrtPriceLowerTick, sqrtPriceUpperTick, err := math.TicksToSqrtPrice(position.LowerTick, position.UpperTick, concentratedPoolExtension.GetPrecisionFactorAtPriceOne())
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		asset0, asset1 := concentratedPoolExtension.CalcActualAmounts(sdkCtx, position.LowerTick, position.UpperTick, sqrtPriceLowerTick, sqrtPriceUpperTick, position.Liquidity)
```
Comment (p0mvn) [inline x/concentrated-liquidity/grpc_query.go:101]: Consider encapsulating this into a function, there are a few other places where we can reuse this

File: x/concentrated-liquidity/grpc_query.go [hunk 3]
```diff
@@ -77,8 +78,37 @@ func (q Querier) UserPositions(ctx context.Context, req *clquery.QueryUserPositi
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
+	var positions []model.PositionWithUnderlyingAssetBreakdown
```
Comment (p0mvn) [inline x/concentrated-liquidity/grpc_query.go:81]: nit: this can be pre-allocated

File: x/concentrated-liquidity/grpc_query.go (hunk 2 above)
Comment (p0mvn) [inline x/concentrated-liquidity/grpc_query.go:101]: Outside of the context of this PR but commenting for visibility: noticing this function is missing tests. I also observe some rounding behavior that would be great to get reviewed separately

File: x/concentrated-liquidity/grpc_query.go (hunk 3 above)
Comment (p0mvn) [inline x/concentrated-liquidity/grpc_query.go:81]: Please cover this logic with some tests

**Review by p0mvn — APPROVED** [review]
Nice work on getting this out so quickly! Had a couple of non-blocking comments and requested tests

File: x/concentrated-liquidity/grpc_query.go (hunk 3 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/grpc_query.go]: Please lmk if this is sufficient! https://github.com/osmosis-labs/osmosis/pull/4637/commits/c3290f73de27ad84ca8bcc04d5de5072fa122628

File: x/concentrated-liquidity/grpc_query.go (hunk 2 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/grpc_query.go]: Created an issue here https://github.com/osmosis-labs/osmosis/issues/4644

File: x/concentrated-liquidity/position.go [hunk 4]
```diff
@@ -150,3 +151,15 @@ func (k Keeper) CreateFullRangePosition(ctx sdk.Context, concentratedPool types.
 
 	return amount0, amount1, liquidity, joinTime, nil
 }
+
+func CalculateUnderlyingAssetsFromPosition(ctx sdk.Context, position model.Position, pool types.ConcentratedPoolExtension) (sdk.Dec, sdk.Dec, error) {
```
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:155]: nit: we can define this on the pool struct as a method instead of passing in the pool as param

**Review by p0mvn — APPROVED** [review]

**Pull request merged**